rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    require_session_mfa: on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  require_session_mfa: on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  role_options:
+    require_session_mfa: on | hardware_key | hardware_key_touch
+```
+
+- `on`: Current per-session MFA functionality - user's are required to pass an MFA challenge with a registered MFA device in order to start new SSH|Kubernetes|DB|Desktop sessions. Non-session requests, and app-session requests are not impacted.
+- `hardware_key`: Essentially combines `require_session_mfa: on` with `private_key_policy: true`. MFA tap is only required for session requests, but all Teleport API requests require certificates backed by a hardware private key.
+- `hardware_key_touch`: This is the same as `private_key_policy: hardware_key_touch`. Per-session MFA is disabled in favor of touch checks directly on the hardware key the user used to login.
+
+Note: From a product perspective, we have decided to add the new `require_session_mfa` options and omit the new `private_key_policy` option to reduce configuration knobs. Since the underlying implementation will be completely separate from per-session MFA, this RFD will continue to refer to `private_key_policy` to make reasoning about the proposed changes easier to understand.
+
+#### Enforcement - Attestation
+
+In order to enforce the user private key usage, we need to take a certificate's public key and tie it back to a trusted hardware device, which can be done with attesatation, as explained [above](#attestation).
+
+First, we need to get attestation data from the user's hardware key during login, which we will do with the new rpc `AttestHardwarePrivateKey`.
+
+```proto
+service AuthService {
+  rpc AttestHardwarePrivateKey(AttestHardwarePrivateKeyRequest) returns (AttestHardwarePrivateKeyResponse);
+}
+
+message AttestHardwarePrivateKeyRequest {
+  YubikeyPIVAttestationRequest yubikey_attestation_request = 1;
+  // We may add non-yubikey and non-piv options in the future
+}
+
+// Data used to attest a slot - https://pkg.go.dev/github.com/go-piv/piv-go@v1.10.0/piv#Verify
+message YubikeyPIVAttestationRequest {
+  bytes slot_cert = 1;
+  bytes attestation_cert = 2;
+}
+
+message AttestHardwarePrivateKeyResponse { }
+```
+
+In addition to verifying the certificate chain for a user's hardware private key to a trusted hardware key manufacturer, the resulting attestation object will provide information about the private key, including;
+ - Device information, including serial number, model, version 
+ - Configured Touch (And PIN) Policies if any
+
+This attestation data will be check against server settings. If the attestation is valid, it will be stored in the backend in `/hardware_key_attestations/<public_key_der>`.
```

Comment:
Is DER a stable format, or are we relying on the hardware key to always generate the same DER?

Is `/hardware_key_attestations/SHA256(public_key_der)` viable as a storage key? Same practical effect, but bounded size.

Reply:
No DER is not viable, SHA256 is a much better idea, thanks!